### PR TITLE
Checking the length of the name field

### PR DESCRIPTION
**Pull request by melnelen**

### CODE REVIEW

File: FocusOn/View/TodayView.swift [hunk 1]
```diff
@@ -78,61 +95,93 @@ struct TodayView: View {
                                     .foregroundColor(tasksAreCompleted[2] ? Color("SuccessColor") : .black)
                             }
                     }
+                    .alert(isPresented: $showAlert) {
+                        Alert(title: Text("Oops 🙊"),
+                              message: Text("Please, make sure that the name of your third task is at least 3 characters long"),
+                              dismissButton: .default(Text("OK")))
+                    }
```
Comment (cyril-lepottier) [inline FocusOn/View/TodayView.swift]: Warning, code duplication

File: FocusOn/View/TodayView.swift [hunk 2]
```diff
@@ -78,61 +95,93 @@ struct TodayView: View {
...
-        // link view goal to the viewModel goal
-        todayGoal = viewModel.todayGoal
-
-        // add the goal to the list of goals
-        viewModel.addGoal(name: goalText)
-
-        // hide keyboard
-        UIApplication.shared.endEditing()
+        // check that the name of the goal is at least 3 characters long
+        nameIsLongEnough = checkLength(of: goalText)
```
Comment (cyril-lepottier) [inline FocusOn/View/TodayView.swift]: As spoken, ideally, the check would be done by the service or in `Goal` itself with some unit tests attached.

Comment (cyril-lepottier) [inline FocusOn/View/TodayView.swift]: As spoken, ideally, the check would be done by the service or in `Goal` itself with some unit tests attached.

**Review by cyril-lepottier — APPROVED** [review]

File: FocusOn/View/TodayView.swift (hunk 2 above)
Comment (melnelen) [reply FocusOn/View/TodayView.swift]: Fixed

Comment (melnelen) [reply FocusOn/View/TodayView.swift]: Fixed

**Comment by melnelen** [comment]
The errors are now thrown by the data service and handled in the view by showing an alert.
Tests have been added.

File: FocusOnTests/TodayViewModelTests.swift [hunk 3]
```diff
@@ -24,41 +24,153 @@ class TodayViewModelTests: XCTestCase {
...
+        }
+        // Then
+        XCTAssert(error == "The name is too short!", "Could not catch the error for a name that is too short when adding a goal")
+    }
+
     func test_TodayViewModel_addGoal() {
         // Given
         // When
-        sut.addGoal(name: "Create FocusOn app")
+        do { try sut.addGoal(name: "Create FocusOn app") } catch { }
```
Comment (cyril-lepottier) [inline FocusOnTests/TodayViewModelTests.swift]: You can also add `throws` on the function:
```
func test_TodayViewModel_updateTask() throws {
```

File: FocusOn/Services/MockDataService.swift [hunk 4]
```diff
@@ -17,16 +17,28 @@ class MockDataService: DataServiceProtocol {
...
+        guard goal.name.count >= 3 else { throw NameLengthError.short }
+
         savedGoals.append(goal)
     }
 
-    func updateGoal(goal: Goal, name: String, isCompleted: Bool) {
+    func updateGoal(goal: Goal, name: String, isCompleted: Bool) throws {
+        // check that the name of the goal is at least 3 characters long
+        guard goal.name.count > 0 else { throw NameLengthError.empty }
+        guard goal.name.count >= 3 else { throw NameLengthError.short }
```
Comment (cyril-lepottier) [inline FocusOn/Services/MockDataService.swift]: Minor, but there is a little code duplication here

**Review by cyril-lepottier — APPROVED** [review]
Nice update :)

**Pull request merged**